aws/resource_aws_s3_bucket_public_access_block.go

```diff
@@ -0,0 +1,183 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/s3control"
+	"github.com/hashicorp/terraform/helper/resource"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsS3BucketPublicAccessBlock() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsS3BucketPublicAccessBlockCreate,
+		Read:   resourceAwsS3BucketPublicAccessBlockRead,
+		Update: resourceAwsS3BucketPublicAccessBlockUpdate,
+		Delete: resourceAwsS3BucketPublicAccessBlockDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"bucket": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"block_public_acls": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+
+			"block_public_policy": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+
+			"ignore_public_acls": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+
+			"restrict_public_buckets": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+		},
+	}
+}
+
+func resourceAwsS3BucketPublicAccessBlockCreate(d *schema.ResourceData, meta interface{}) error {
+	s3conn := meta.(*AWSClient).s3conn
+	bucket := d.Get("bucket").(string)
+
+	input := &s3.PutPublicAccessBlockInput{
+		Bucket: aws.String(bucket),
+		PublicAccessBlockConfiguration: &s3.PublicAccessBlockConfiguration{
+			BlockPublicAcls:       aws.Bool(d.Get("block_public_acls").(bool)),
+			BlockPublicPolicy:     aws.Bool(d.Get("block_public_policy").(bool)),
+			IgnorePublicAcls:      aws.Bool(d.Get("ignore_public_acls").(bool)),
+			RestrictPublicBuckets: aws.Bool(d.Get("restrict_public_buckets").(bool)),
+		},
+	}
+
+	log.Printf("[DEBUG] S3 bucket: %s, public access block: %v", bucket, input.PublicAccessBlockConfiguration)
+	_, err := s3conn.PutPublicAccessBlock(input)
+	if err != nil {
+		return fmt.Errorf("error creating public access block policy for S3 bucket (%s): %s", bucket, err)
+	}
+
+	d.SetId(bucket)
+	return resourceAwsS3BucketPublicAccessBlockRead(d, meta)
+}
+
+func resourceAwsS3BucketPublicAccessBlockRead(d *schema.ResourceData, meta interface{}) error {
+	s3conn := meta.(*AWSClient).s3conn
+
+	input := &s3.GetPublicAccessBlockInput{
+		Bucket: aws.String(d.Id()),
+	}
+
+	// Retry for eventual consistency on creation
+	var output *s3.GetPublicAccessBlockOutput
+	err := resource.Retry(1*time.Minute, func() *resource.RetryError {
+		var err error
+		output, err = s3conn.GetPublicAccessBlock(input)
+
+		if d.IsNewResource() && (isAWSErr(err, s3control.ErrCodeNoSuchPublicAccessBlockConfiguration, "") ||
+			isAWSErr(err, s3.ErrCodeNoSuchBucket, "")) {
```

Comment:
> The only bits missing from this pull request are around retrying for s3.ErrCodeNoSuchBucket on Create and d.IsNewResource() in Read because S3 eventual consistency is a pain.

@bflad was this what you meant?